src/serac/physics/utilities/variational_form/weak_form.hpp

```diff
@@ -0,0 +1,472 @@
+      }
+    }
+  }
+
+  /**
+   * @brief The input set of local DOF values (i.e., on the current rank)
+   */
+  mutable mfem::Vector input_L_;
+
+  /**
+   * @brief The output set of local DOF values (i.e., on the current rank)
+   */
+  mutable mfem::Vector output_L_;
+
+  /**
+   * @brief The output set of local DOF values (i.e., on the current rank) from boundary elements
+   */
+  mutable mfem::Vector output_L_boundary_;
+
+  /**
+   * @brief The input set of per-element DOF values
+   */
+  mutable mfem::Vector input_E_;
+
+  /**
+   * @brief The input set of per-boundaryelement DOF values
+   */
+  mutable mfem::Vector input_E_boundary_;
+
+  /**
+   * @brief The output set of per-element DOF values
+   */
+  mutable mfem::Vector output_E_;
+
+  /**
+   * @brief The output set of per-boundary-element DOF values
+   */
+  mutable mfem::Vector output_E_boundary_;
+
+  /**
+   * @brief The set of true DOF values, used as a scratchpad for @p operator()
+   */
+  mutable mfem::Vector my_output_T;
+  /**
+   * @brief A working vector for @p GetGradient
+   */
+  mutable mfem::Vector dummy_;
+
+  /**
+   * @brief Manages DOFs for the test space
+   */
+  mfem::ParFiniteElementSpace* test_space_;
+
+  /**
+   * @brief Manages DOFs for the trial space
+   */
+  mfem::ParFiniteElementSpace* trial_space_;
+
+  /**
+   * @brief The set of true DOF indices to which an essential BC should be applied
+   */
+  mfem::Array<int> ess_tdof_list_;
+
+  /**
+   * @brief Operator that converts true (global) DOF values to local (current rank) DOF values
+   * for the test space
+   */
+  const mfem::Operator* P_test_;
+
+  /**
+   * @brief Operator that converts local (current rank) DOF values to per-element DOF values
+   * for the test space
+   */
+  const mfem::Operator* G_test_;
+
+  /**
+   * @brief Operator that converts local (current rank) DOF values to per-boundary element DOF values
+   * for the test space
+   */
+  const mfem::Operator* G_test_boundary_;
+
+  /**
+   * @brief Operator that converts true (global) DOF values to local (current rank) DOF values
+   * for the trial space
+   */
+  const mfem::Operator* P_trial_;
+
+  /**
+   * @brief Operator that converts local (current rank) DOF values to per-element DOF values
+   * for the trial space
+   */
+  const mfem::Operator* G_trial_;
+
+  /**
+   * @brief Operator that converts local (current rank) DOF values to per-boundary element DOF values
+   * for the trial space
+   */
+  const mfem::Operator* G_trial_boundary_;
+
+  /**
+   * @brief The set of domain integrals (spatial_dim == geometric_dim)
+   */
+  std::vector<Integral<test(trial)> > domain_integrals_;
+
+  /**
+   * @brief The set of boundary integral (spatial_dim > geometric_dim)
+   */
+  std::vector<Integral<test(trial)> > boundary_integrals_;
+
+  // simplex elements are currently not supported;
+  static constexpr mfem::Element::Type supported_types[4] = {mfem::Element::POINT, mfem::Element::SEGMENT,
```

Comment:
Trailing underscore

Reply:
Actually, this could probably be a static variable outside of this class.

Reply:
```suggestion
  static constexpr mfem::Element::Type supported_types_[4] = {mfem::Element::POINT, mfem::Element::SEGMENT,
```